### PR TITLE
Improve performance by not extracting compressed image data if retainImageContent was set to false

### PR DESCRIPTION
**Pull request by se-ti**
One of previous pull requests allowed not to decompress image data, if the user has set retainImageContent flag to false.
This pull request suggests even not to extract compressed image data, if this flag was set to false.

Minor optimization: extracting stream data tries to use header information about stream's length to speed preg_match up.

This PR improves memory allocation, achieved with [pull 582](https://github.com/smalot/pdfparser/pull/582),
from 45M for the first launch, 25M for the best subsequent one 
to 43M for the first launch, 18M for the best subsequent.

https://westra.ru/reports/kavkaz/danilina-1el2-arhyz-2021.pdf
```php
$lim = '128M';
ini_set("memory_limit", $lim);
include 'alt_autoload.php-dist';

$path = 'danilina-1el2-arhyz-2021.pdf';

$config = new \Smalot\PdfParser\Config();
$config->setRetainImageContent(false);
$parser = new \Smalot\PdfParser\Parser([], $config);

$content = file_get_contents($path);
$pdf = $parser->parseContent($content);

$text = $pdf->getText();

echo ini_get("memory_limit")."\n";
echo $text;
```

### CODE REVIEW

**Comment by se-ti** [comment]
Sorry, i don't have PHPStan installed and set up.
What is the problem with it?

How can i interpret its output?
```
Loaded config default from "/home/runner/work/pdfparser/pdfparser/.php-cs-fixer.php".
................F................................................ 65 / 72 ( 90%)
.......                                                           72 / 72 (100%)
Legend: .-no changes, F-fixed, S-skipped (cached or empty file), I-invalid file syntax (file ignored), E-error
   1) src/Smalot/PdfParser/RawData/RawDataParser.php (modernize_types_casting, method_argument_space, no_unneeded_control_parentheses, no_superfluous_phpdoc_tags, native_function_invocation, yoda_style, phpdoc_align)
```

**Comment by k00ni** [comment]
> How can i interpret its output?

PHP-CS-Fixer complains the file. Give me a day or two, if you don't want to install PHP-CS-Fixer locally (to fix it).

File: src/Smalot/PdfParser/RawData/RawDataParser.php [hunk 1]
```diff
@@ -758,15 +761,20 @@ protected function getRawObject(string $pdfData, int $offset = 0): array
                     $offset += 6;
                     if (1 == preg_match('/^([\r]?[\n])/isU', substr($pdfData, $offset, 4), $matches)) {
                         $offset += \strlen($matches[0]);
+
+                        $streamLen = (int) $this->getHeaderValue($headerDic, 'Length', 'numeric', 0);
```
Comment (k00ni) [inline src/Smalot/PdfParser/RawData/RawDataParser.php:765]: Please add one or more sentences about this new code.

**Review by k00ni — CHANGES_REQUESTED** [review]
_Sorry, I didn't had the time to answer properly. Now I have._

First of all, thank you again for your work @se-ti.

I pushed a few changes fixing coding style issues (the ones you mentioned) and made the new method `getRawObject` private.

Further remarks in code.

File: src/Smalot/PdfParser/RawData/RawDataParser.php [hunk 2]
```diff
@@ -758,15 +761,20 @@ protected function getRawObject(string $pdfData, int $offset = 0): array
...
+                        $skip = !$this->config->getRetainImageContent() && 'XObject' == $this->getHeaderValue($headerDic, 'Type', '/') && 'Image' == $this->getHeaderValue($headerDic, 'Subtype', '/');
```
Comment (k00ni) [inline src/Smalot/PdfParser/RawData/RawDataParser.php:766]: Replace the first part (`!$this->...`) with something like `false === `. It is more readable and will also improve maintenance in the future.

File: src/Smalot/PdfParser/RawData/RawDataParser.php (hunk 1 above)
Comment (se-ti) [reply src/Smalot/PdfParser/RawData/RawDataParser.php:765]: If you look through PDFs, you'll find that stream blocks (stream-endstream) are anticipated with header section (<< >>) with a field Length, that contains length of this stream.

This fact is already used in protected function decodeStream
https://github.com/smalot/pdfparser/blob/65bdad55246ec230a055fbda501c6fed592c1942/src/Smalot/PdfParser/RawData/RawDataParser.php#L100

So we can extract this length, add it to preg_match's offset and:
-- help preg_match to test less data, i.e. work faster
-- prevent (not very probable) false match, when compressed data contains 'endstream ' by itself

File: src/Smalot/PdfParser/RawData/RawDataParser.php (hunk 2 above)
Comment (se-ti) [reply src/Smalot/PdfParser/RawData/RawDataParser.php]: done

File: src/Smalot/PdfParser/RawData/RawDataParser.php [hunk 3]
```diff
@@ -796,6 +804,38 @@ protected function getRawObject(string $pdfData, int $offset = 0): array
...
+                \is_array($val)
+                && 3 == \count($val)
+                && '/' == $val[0]
+                && $val[1] == $key
+                && isset($headerDic[$i + 1])
+            ) {
+                return \is_array($headerDic[$i + 1]) && 1 < \count($headerDic[$i + 1]) && $type == $headerDic[$i + 1][0]
+                    ? $headerDic[$i + 1][1]
+                    : $default;
+            }
```
[the remark this answers is not in the document]
Comment (se-ti) [reply src/Smalot/PdfParser/RawData/RawDataParser.php]: This is a utility function, that provides access to object header's fields (obj << YYY >> part ), like Header::get('...')->getContent() do.
The only difference -- it can be used during the parsing process, when no Smalot\PdfParser\Header objects are created yet.

It recieves dictionary of header fields, as it is returned by RawDataParser::getRawObject,
iterates over it, searching for section of type '/' whith requested key.
If such a section is found, it tries to receive it's value (next object in dictionary), returning it, if it matches requested type, or default value otherwise.

This function can simplify index arithmetic here: 
https://github.com/smalot/pdfparser/blob/65bdad55246ec230a055fbda501c6fed592c1942/src/Smalot/PdfParser/RawData/RawDataParser.php#L262
It's just $valid_crs = 'XRef' == $this->getHeaderValue($sdik, 'Type', '/');
and several ifs below this line;

Here:
https://github.com/smalot/pdfparser/blob/65bdad55246ec230a055fbda501c6fed592c1942/src/Smalot/PdfParser/RawData/RawDataParser.php#L100
And there:
https://github.com/smalot/pdfparser/blob/65bdad55246ec230a055fbda501c6fed592c1942/src/Smalot/PdfParser/RawData/RawDataParser.php#L107

Answering your questions:

1. I don't see any more edge cases, we have to test, since i'm testing correctness of inner structure of an object, returned by RawDataParser::getRawObject (those is_array, count() >= X). Even this may be a bit excessive, but let it be :)
 
2. It's just a separation of responsibility, nothing more. if ensures, that desired field is found and it has some value, ternary operator ensures that value object is correct and has desired type.
It could be reduced to one if, if you insist.

**Comment by se-ti** [comment]
>  I pushed a few changes fixing coding style issues...
Thanks a lot, @k00ni !

File: src/Smalot/PdfParser/RawData/RawDataParser.php [hunk 4]
```diff
@@ -796,6 +804,39 @@ protected function getRawObject(string $pdfData, int $offset = 0): array
...
+     *
+     * @param string            $key     header's section name
+     * @param string            $type    type of the section (i.e. 'numeric', '/', '<<', etc.)
+     * @param string|array|null $default default value for header's section
+     *
+     * @return string|array|null value of obj header's section, or default value if none found, or its type doesn't match $type param
+     */
+    private function getHeaderValue(?array $headerDic, string $key, string $type, $default = '')
+    {
+        if (!\is_array($headerDic)) {
```
Comment (k00ni) [inline src/Smalot/PdfParser/RawData/RawDataParser.php:818]:
```suggestion
        if (false === \is_array($headerDic)) {
```

Does it make sense to you to expand this by including a check that `$headerDic` is not empty?

Comment (se-ti) [reply src/Smalot/PdfParser/RawData/RawDataParser.php]: I don't see any sense to add this empty check. 
If it's empty, foreach will just do nothing and function will return $default at the last line.

P.S. A comment in [doc](https://www.php.net/manual/en/language.operators.logical.php#90386):
> Evaluation of logical expressions is stopped as soon as the result is known.

It's too powerful way of optimization to ignore it

**Comment by se-ti** [comment]
I think, I have added all the requested changes.
Should i do anything else?
Any other requests?
@k00ni

**Comment by se-ti** [comment]
Аny news, @k00ni?

**Review by k00ni — APPROVED** [review]
@se-ti I am very busy currently, but I will try to squeeze in a few minutes here and there. Thank you for bearing with me.

I added a few comments to the code based on your latest feedback. My knowledge in this topic is limited so I will have to trust you on this one. It looks clean and results speak for themselves.

I will leave this open for a few days to give others a chance to object.

**Comment by se-ti** [comment]
Thank You!
Waiting for this PR to be merged, to suggest another one.
